File: backend/app/api/meta.py
"""筛选元数据接口 —— **所有者：倪嵩**。

见 docs/00_协作规范.md §3 文件所有权表。莫炜钧在骨架阶段先把文件建好
并在 main.py 注册，倪嵩直接实现本文件，**不要改 main.py**。

────────────────────────────────────────────────────────────────
倪嵩：契约见《任务执行书》§4.5。

    GET /api/meta/filters

返回当前数据库**真实可用**的：
    years / source_provinces / edu_levels / categories / batches

要点（附录 B 第四条）：
- 从数据库真实值生成，不要硬编码全国 category / batch。
- category / batch 是来源原始口径，不要自行造全国统一映射。
- 可以去空值、合理排序，但不要改数据库。

参考：school_admission 里 2023—2025 三年、30 个生源省、
category 17 个值、batch 174 个值；college.edu_level 只有两个值。
────────────────────────────────────────────────────────────────
"""

from fastapi import APIRouter

from app.services import admission_service
from app.warnings import SOURCE_VOCABULARY, dedupe

router = APIRouter(tags=["meta"])


@router.get("/meta/filters")
def meta_filters():
    """返回筛选器元数据：年份、生源省、办学层次、科类、批次。

    category / batch 为来源数据原始口径，未做全国统一标准化。
    """
    meta = admission_service.filter_meta()

    # 涉及 category / batch，加来源口径警告
    warnings = dedupe([SOURCE_VOCABULARY])

    return {
        "data": meta,
        "warnings": warnings,
    }

File: frontend/src/types/meta.ts
export interface FilterMeta {
  years: number[];
  source_provinces: string[];
  edu_levels: string[];
  categories: string[];
  batches: string[];
}

export interface FilterMetaResponse {
  data: FilterMeta;
  warnings: string[];
}

/** SchoolAdmission 中至少有一条有效 min_rank 的真实考试上下文。 */
export interface CandidateProfileContext {
  source_province: string;
  year: number;
  category: string;
  batch: string | null;
}

export interface CandidateProfileContextsResponse {
  data: {
    contexts: CandidateProfileContext[];
  };
  warnings: string[];
}

export interface ScoreRangeContext {
  source_province: string;
  year: number;
  category: string;
  valid_score_count: number;
  ambiguous_score_count: number;
  min_score: number;
  max_score: number;
}

export interface ScoreRangeContextsResponse {
  data: {
    contexts: ScoreRangeContext[];
  };
  warnings: string[];
}

export type ScoreResolutionStatus =
  | "resolved"
  | "ambiguous"
  | "not_found"
  | "unsupported_context";

export interface ScoreResolution {
  source_province: string;
  year: number;
  category: string;
  score: number;
  status: ScoreResolutionStatus;
  resolved_rank: number | null;
  matching_row_count: number;
  distinct_rank_count: number;
}

export interface ScoreResolutionResponse {
  data: ScoreResolution;
  warnings: string[];
}
